### ISSUE TITLE
Masked scatter

### ISSUE THREAD

**Issue by RaulPPelaez** [rusty1s/pytorch_scatter#380]
Hi! Thanks for this wonderful library. I have a feature request :P
In trying to CUDA graph a model I end up with an array of indexes that must have a static shape but in which only some indexes are valid (say >=0). 
As of now I have to provide scatter with only the valid indexes: 
```python
index = torch.Tensor([0,1,0,1,0,-1,-1,-1])
src = torch.Tensor([1,1])
mask = index>=0
index = index[mask]
out = scatter(src, index, dim_size=2)
```
I would like to be able to do this instead:
```python
index = torch.Tensor([0,1,0,1,0,-1,-1,-1])
src = torch.Tensor([1,1])
mask = index>=0
out = masked_scatter(src, index, mask)
```
The hard part is that I would expect this to be compatible with CUDA graphs (as scatter is right now), which means that masked_scatter could not simply do ```index = index[mask]```, rather letting each kernel ignore masked indexes.
Would you consider adding this functionality?
Maybe there is some evident workaround I am missing?

Skimming through the code it seems like the forward pass would be fairly simple to implement. However, for the backwards torch::gather is used, for which there is no masked_gather AFAIK.

Thanks!

EDIT: A possible workaround is to replace the -1 in index by src.numel() and add an extra ghost value to src:
```python
index = torch.Tensor([0,1,0,1,0,2,2,2])
src = torch.Tensor([1,1])
src_extended = torch.cat((src,torch.zeros(1))
out = scatter(src, index, dim_size=3)[:-1]
```
Since scatter uses atomic, this can be slow if there are a lot of masked values...

**Comment by rusty1s** [comment]
Sorry, I somehow missed this issue :( I know that my answer comes rather late. The two workarounds you posted are what I would have suggested as well. Currently, there does not exist a plan to add masking capabilities directly into scatter ops.
